_projects/nextgen_action_system.md

```diff
@@ -0,0 +1,42 @@
+---
+name: "Next Generation Action System"
+desc: "Actions in coala should be improved to allow for more dynamic and suiting actions"
+requirements:
+ - "The applicant must understand how results and actions work in coala."
+ - "The proposal needs to contain a clear concept."
+difficulty: "high"
+mentors:
+ - Makman2
+ - jayvdb
+initiatives:
+ - GSoC
+tags:
+ - core
+---
+
+Currently, `Result` objects are passed to action objects as parameters when coala tries to apply an action. This is a bit restrictive as this means that coala needs to have a global set of actions that it tries to apply on all `Result` objects. Sure, coala checks if an action `is_applicable()` before it tries to apply it but it needs to do that for every `Result` that every bear ever generates.
+
+We need a way allow results to define actions by themselves, like `result = Result(actions=[ApplyPatchAction, ...]). Here is the benefit of such an approach:
+* Bears can specify their own actions, for example the GitCommitBear would highly benefit from this (actions could be added for amending commits). Then, the bear writers could get much more creative and generate even more useful bears.
```

Comment:
why isnt amending a commit applicable to `ApplyPatchAction` ?

The problem is the 'apply' isnt to a traditional file, however it is a patchable file format; it is the storage of the amended version which is different.

The problem becomes quite different if we have 'virtual files', e.g. a component (not yet defined) creates a `.COMMIT_MSG` from the commit message, which is then validated by normal bears, including SpaceConsistencyBear, and then when modified it is patched back into the commit.

Reply:
Okay, I'll keep this in mind when creating the action. :-)

Reply:
> why isnt amending a commit applicable to ApplyPatchAction ?

I actually though about just changing the commit action. A file change + amend might be implemented with a combination of ApplyPatchAction and AmmendAction (using a ChainAction).